File: python/yacht/yacht.py
# Score categories

from abc import ABC, abstractmethod


class Category(ABC):
    @staticmethod
    @abstractmethod
    def check(dice):
        pass

    @staticmethod
    @abstractmethod
    def score(dice):
        pass


class ONES(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 1, dice))


class TWOS(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 2, dice))


class THREES(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 3, dice))


class FOURS(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 4, dice))


class FIVES(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 5, dice))


class SIXES(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(filter(lambda x: x == 6, dice))


class FULL_HOUSE(Category):
    def check(dice):
        return dice[0] == dice[1] and dice[3] == dice[4] and dice[0] != dice[4] and (dice[2] == dice[1] or dice[2] == dice[3])

    def score(dice):
        return sum(dice)


class FOUR_OF_A_KIND(Category):
    def check(dice):
        return dice[0] == dice[3] or dice[1] == dice[4]

    def score(dice):
        return dice[1] * 4


class LITTLE_STRAIGHT(Category):
    def check(dice):
        return all(map(lambda x: x in dice, range(1, 6)))

    def score(dice):
        return 30


class BIG_STRAIGHT(Category):
    def check(dice):
        return all(map(lambda x: x in dice, range(2, 7)))

    def score(dice):
        return 30


class CHOICE(Category):
    def check(dice):
        return True

    def score(dice):
        return sum(dice)


class YACHT(Category):
    def check(dice):
        return dice[0] == dice[4]

    def score(dice):
        return 50


def score(dice, category):
    return category.score(dice) if category.check(sorted(dice)) else 0

File: java/yacht/src/main/java/Yacht.java
import java.util.Arrays;
import java.util.stream.IntStream;

class Yacht {
  int[] dice;
  YachtCategory yachtCategory;

  Yacht(int[] dice, YachtCategory yachtCategory) {
    this.dice = Arrays.stream(dice).sorted().toArray();
    this.yachtCategory = yachtCategory;
  }

  int score() {
    if (yachtCategory == YachtCategory.ONES) {
      return Arrays.stream(dice).filter(x -> x == 1).sum();
    }
    if (yachtCategory == YachtCategory.TWOS) {
      return Arrays.stream(dice).filter(x -> x == 2).sum();
    }
    if (yachtCategory == YachtCategory.THREES) {
      return Arrays.stream(dice).filter(x -> x == 3).sum();
    }
    if (yachtCategory == YachtCategory.FOURS) {
      return Arrays.stream(dice).filter(x -> x == 4).sum();
    }
    if (yachtCategory == YachtCategory.FIVES) {
      return Arrays.stream(dice).filter(x -> x == 5).sum();
    }
    if (yachtCategory == YachtCategory.SIXES) {
      return Arrays.stream(dice).filter(x -> x == 6).sum();
    }
    if (yachtCategory == YachtCategory.FULL_HOUSE && isFullHouse()) {
      return Arrays.stream(dice).sum();
    }
    if (yachtCategory == YachtCategory.FOUR_OF_A_KIND && isFourOfAKind()) {
      return dice[1] * 4;
    }
    if (yachtCategory == YachtCategory.LITTLE_STRAIGHT && isLittleStraight()) {
      return 30;
    }
    if (yachtCategory == YachtCategory.BIG_STRAIGHT && isBigStraight()) {
      return 30;
    }
    if (yachtCategory == YachtCategory.CHOICE) {
      return Arrays.stream(dice).sum();
    }
    if (yachtCategory == YachtCategory.YACHT && isYacht()) {
      return 50;
    }

    return 0;
  }

  boolean isFullHouse() {
    return dice[0] == dice[1]
        && dice[3] == dice[4]
        && dice[0] != dice[4]
        && (dice[2] == dice[1] || dice[2] == dice[3]);
  }

  boolean isFourOfAKind() {
    return dice[0] == dice[3] || dice[1] == dice[4];
  }

  boolean isLittleStraight() {
    return IntStream.range(0, 5).allMatch(i -> dice[i] == i + 1);
  }

  boolean isBigStraight() {
    return IntStream.range(0, 5).allMatch(i -> dice[i] == i + 2);
  }

  boolean isYacht() {
    return dice[0] == dice[4];
  }
}
